### ISSUE TITLE
[BUG] No type safety when passing argument to @qml.device() decorated function 

### ISSUE THREAD

**Issue by kharazity** [PennyLaneAI/pennylane#1677]
### Expected behavior

It would be really nice if I could combine circuits in an obvious way, but the lack of type safety prevents me from doing so whenever the function is decorated with `@qml.qnode(device)`. I really simple example is preparing a VQE circuit from a UCCSD ansatz plus grouped measurement. 

I understand that using the built-in methods should take care of most of this for me, but for the purposes of what I am trying to do that simply won't do.

As a simple example, consider H2 with qubit-wise commuting groups.

```
import pennylane as qml
from pennylane import qchem
import numpy as np
from functools import partial
from pennylane.templates import UCCSD

#MOLECULE
symbols = ["H", "H"]
coordinates = np.array([0.0, 0.0, -0.6614, 0.0, 0.0, 0.6614])
h2_ham, n_qubits = qchem.molecular_hamiltonian(symbols, coordinates)
n_electrons = 2

groupings, coeffs = qml.grouping.group_observables(h2_ham.terms[1], h2_ham.terms[0], grouping_type = 'qwc')
dev_noisy = qml.device('default.mixed', wires = n_qubits, shots = 1000)

#prepare UCCSD ansatz
singles, doubles = qchem.excitations(n_electrons, n_qubits)
s_wires, d_wires = qchem.excitations_to_wires(singles, doubles)
ref_state = qchem.hf_state(n_electrons, n_qubits)
ansatz = partial(UCCSD, init_state = ref_state, s_wires = s_wires, d_wires = d_wires)

@qml.qnode(dev_noisy)
def VQE_circuit(params, group, n_qubits):
    print(type(group))
    qml.apply(ansatz(params, wires = range(n_qubits)))
    #FAILS HERE, tensor has no attribute wire
    rotations = qml.grouping.diagonalize_qwc_pauli_words(group)[0]
    for rotation in rotations:
        qml.apply(rotation)
    return qml.sample(wires = range(n_qubits))
params = np.random.randn(len(s_wires) + len(d_wires))

circuits = [VQE_circuit(params, group, n_qubits) for group in groupings]
```

### Actual behavior

However, the lack of type safety prevents me from doing so. 

Prior to passing `group` to `VQE_circuit`,
`type(group) = <class 'list'>` with a particular object being something like: `[PauliY(wires=[0]) @ PauliY(wires=[1]) @ PauliX(wires=[2]) @ PauliX(wires=[3])]`.

However, once it's in the body of the function:
`type(group) = <class 'pennylane.numpy.tensor.tensor'>`
which poses a problem since that class doesn't interface with `diagonalize_qwc_pauli_words` bc `pennylane.numpy.tensor.tensor` objects don't have `wires`. 

Note that this all works perfectly fine once the decorator `@qml.device` is removed.

### System information

```shell
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Name: PennyLane
Version: 0.17.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /home/kharazi/.local/lib/python3.9/site-packages
Requires: networkx, toml, semantic-version, appdirs, autograd, scipy, autoray, numpy
Required-by: VQELib, PennyLane-Qchem
Platform info:           Linux-5.11.0-34-generic-x86_64-with-glibc2.33
Python version:          3.9.5
Numpy version:           1.21.2
Scipy version:           1.7.1
Installed devices:
- default.gaussian (PennyLane-0.17.0)
- default.mixed (PennyLane-0.17.0)
- default.qubit (PennyLane-0.17.0)
- default.qubit.autograd (PennyLane-0.17.0)
- default.qubit.jax (PennyLane-0.17.0)
- default.qubit.tf (PennyLane-0.17.0)
- default.tensor (PennyLane-0.17.0)
- default.tensor.tf (PennyLane-0.17.0)
```

### 

- [x] I have searched exisisting GitHub issues to make sure the issue does not already exist.

**Comment by josh146** [comment]
Hi @kharazity! Thanks for catching this bug.

After investigating, this bug is in fact related to #1459. The root cause is that, when using `autograd`, all positional arguments passed to a QNode will be wrapped implicitly as a `np.tensor` in order for autodifferentiation to work.

This is something we are working on fixing on our end, and is very high priority!

In the meantime, a workaround is to always pass non-array arguments as *keyword* arguments. For example, the following works for me:

```python
@qml.qnode(dev_noisy)
def VQE_circuit(params, group, n_qubits):
    ansatz(params, wires=range(n_qubits))
    rotations = qml.grouping.diagonalize_qwc_pauli_words(group)[0]
    return qml.sample(wires=range(n_qubits))

params = np.random.randn(len(s_wires) + len(d_wires))

for group in groupings:
    print(qml.draw(VQE_circuit)(params, group=group, n_qubits=n_qubits))
    print("Result:", VQE_circuit(params, group=group, n_qubits=n_qubits))
```

One other thing is that I removed the `qml.apply()` from inside the QNode. This is because all PennyLane functions apply themselves automatically on instantiation, so the `qml.apply()` is not needed (in fact, it will cause the operations to be applied twice!):

```python
>>> dev = qml.device("default.qubit", wires=1)
>>> @qml.qnode(dev)
... def circuit(x):
...     qml.apply(qml.RX(x, wires=0))
...     return qml.expval(qml.PauliZ(0))
>>> print(qml.draw(circuit)(0.5))
 0: ──RX(0.5)──RX(0.5)──┤ ⟨Z⟩
```

`qml.apply()` however is useful when you want to apply an operation that was previously instantiated outside the QNode, or if you want to repeat the same operation multiple times:

```python
>>> dev = qml.device("default.qubit", wires=1)
>>> op = qml.RX(0.6, wires=0)
>>> @qml.qnode(dev)
... def circuit():
...     qml.apply(op)
...     return qml.expval(qml.PauliZ(0))
>>> print(qml.draw(circuit)())
0: ──RX(0.6)──┤ ⟨Z⟩
```

**Comment by kharazity** [comment]
Thanks for getting back! I'm new to pennylane, the tips RE `qml.apply()` were very helpful. The keyword arguments seems a good workaround! 

Thanks!

**Comment by josh146** [comment]
No worries @kharazity! Since this is a duplicate of an existing bug, I will close this in favour of #1459. However, don't hesitate to open a new issue if you discover any other bugs!

In addition, we also have a PennyLane discussion forum, where you can ask general code and QML related questions: https://discuss.pennylane.ai

**Comment by kharazity** [comment]
Just curious, does the code you posted run as is? I am currently getting an `ValueError: Observable Identity(wires=[0]) does not have a measurement type specified.`. When running that snippet. This behavior is perplexing to me.

**Comment by josh146** [comment]
Hi @kharazity, that is very odd, it does run for me 🤔 Here is the full code snippet:

```python
import pennylane as qml
from pennylane import qchem
import numpy as np
from functools import partial
from pennylane.templates import UCCSD

#MOLECULE
symbols = ["H", "H"]
coordinates = np.array([0.0, 0.0, -0.6614, 0.0, 0.0, 0.6614])
h2_ham, n_qubits = qchem.molecular_hamiltonian(symbols, coordinates)
n_electrons = 2

groupings, coeffs = qml.grouping.group_observables(h2_ham.terms[1], h2_ham.terms[0], grouping_type = 'qwc')
dev_noisy = qml.device('default.mixed', wires = n_qubits, shots = 1000)

#prepare UCCSD ansatz
singles, doubles = qchem.excitations(n_electrons, n_qubits)
s_wires, d_wires = qchem.excitations_to_wires(singles, doubles)
ref_state = qchem.hf_state(n_electrons, n_qubits)
ansatz = partial(UCCSD, init_state = ref_state, s_wires = s_wires, d_wires = d_wires)

@qml.qnode(dev_noisy)
def VQE_circuit(params, group, n_qubits):
    ansatz(params, wires=range(n_qubits))
    rotations = qml.grouping.diagonalize_qwc_pauli_words(group)[0]
    return qml.sample(wires=range(n_qubits))

params = np.random.randn(len(s_wires) + len(d_wires))

for group in groupings:
    print(qml.draw(VQE_circuit)(params, group=group, n_qubits=n_qubits))
    print("Result:", VQE_circuit(params, group=group, n_qubits=n_qubits))
```

**Comment by kharazity** [comment]
Yeah, just ran that snippet. Same error. What version of pennylane are you using?

Here's the full error message: 
```
ValueError: Observable Identity(wires=[0]) does not have a measurement type specified.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_34521/3633443393.py in <module>
     29 
     30 for group in groupings:
---> 31     print(qml.draw(VQE_circuit)(params, group=group, n_qubits=n_qubits))
     32     print("Result:", VQE_circuit(params, group=group, n_qubits=n_qubits))

~/.local/lib/python3.9/site-packages/pennylane/transforms/draw.py in wrapper(*args, **kwargs)
     89     @wraps(qnode)
     90     def wrapper(*args, **kwargs):
---> 91         qnode.construct(args, kwargs)
     92         _wire_order = wire_order or qnode.device.wires
     93         _wire_order = qml.wires.Wires(_wire_order)

~/.local/lib/python3.9/site-packages/pennylane/qnode.py in construct(self, args, kwargs)
    523 
    524         with self.qtape:
--> 525             self.qfunc_output = self.func(*args, **kwargs)
    526 
    527         if not isinstance(self.qfunc_output, Sequence):

~/.local/lib/python3.9/site-packages/pennylane/tape/tape.py in __exit__(self, exception_type, exception_value, traceback)
    352         try:
    353             super().__exit__(exception_type, exception_value, traceback)
--> 354             self._process_queue()
    355         finally:
    356             QuantumTape._lock.release()

~/.local/lib/python3.9/site-packages/pennylane/tape/tape.py in _process_queue(self)
    452             elif isinstance(obj, qml.operation.Observable) and "owner" not in info:
    453                 print("PENNYLANE INFO: ", info)
--> 454                 raise ValueError(f"Observable {obj} does not have a measurement type specified.")
    455 
    456         self._update()

ValueError: Observable Identity(wires=[0]) does not have a measurement type specified.
```

I went into `/pennylane/tape/tape.py` and printed info. 

```
            elif isinstance(obj, qml.operation.Observable) and "owner" not in info:
                print("PENNYLANE INFO: ", info)
                raise ValueError(f"Observable {obj} does not have a measurement type specified.")
```
which prints `PENNYLANE INFO: {}`

**Comment by kharazity** [comment]
Actually, just upgraded to Pennylane v18. Apparently it fixes that issue at the cost of getting a `ValueError: probabilities are not non-negative` when running this line: `results = [VQE_circuit(params, group = group, n_qubits = n_qubits) for group in groupings]`.

The full error message is: 
```
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_35105/3078197478.py in <module>
     22 
     23 params = np.random.randn(len(s_wires) + len(d_wires))
---> 24 results = [VQE_circuit(params, group = group, n_qubits = n_qubits) for group in groupings]

/tmp/ipykernel_35105/3078197478.py in <listcomp>(.0)
     22 
     23 params = np.random.randn(len(s_wires) + len(d_wires))
---> 24 results = [VQE_circuit(params, group = group, n_qubits = n_qubits) for group in groupings]

~/.local/lib/python3.9/site-packages/pennylane/qnode.py in __call__(self, *args, **kwargs)
    685             res = fn(results)
    686         else:
--> 687             res = self.qtape.execute(device=self.device)
    688 
    689         finite_diff = any(

~/.local/lib/python3.9/site-packages/pennylane/tape/tape.py in execute(self, device, params)
   1320             params = self.get_parameters()
   1321 
-> 1322         return self._execute(params, device=device)
   1323 
   1324     def execute_device(self, params, device):

~/.local/lib/python3.9/site-packages/autograd/tracer.py in f_wrapped(*args, **kwargs)
     46             return new_box(ans, trace, node)
     47         else:
---> 48             return f_raw(*args, **kwargs)
     49     f_wrapped.fun = f_raw
     50     f_wrapped._is_autograd_primitive = True

~/.local/lib/python3.9/site-packages/pennylane/interfaces/autograd.py in _execute(self, params, device)
    163         # evaluate the tape
    164         self.set_parameters(self._all_params_unwrapped, trainable_only=False)
--> 165         res = self.execute_device(params, device=device)
    166         self.set_parameters(self._all_parameter_values, trainable_only=False)
    167 

~/.local/lib/python3.9/site-packages/pennylane/tape/tape.py in execute_device(self, params, device)
   1351 
   1352         if isinstance(device, qml.QubitDevice):
-> 1353             res = device.execute(self)
   1354         else:
   1355             res = device.execute(self.operations, self.observables, {})

~/.local/lib/python3.9/site-packages/pennylane/_qubit_device.py in execute(self, circuit, **kwargs)
    196         # generate computational basis samples
    197         if self.shots is not None or circuit.is_sampled:
--> 198             self._samples = self.generate_samples()
    199 
    200         multiple_sampled_jobs = circuit.is_sampled and self._has_partitioned_shots()

~/.local/lib/python3.9/site-packages/pennylane/_qubit_device.py in generate_samples(self)
    473         rotated_prob = self.analytic_probability()
    474 
--> 475         samples = self.sample_basis_states(number_of_states, rotated_prob)
    476         return QubitDevice.states_to_binary(samples, self.num_wires)
    477 

~/.local/lib/python3.9/site-packages/pennylane/_qubit_device.py in sample_basis_states(self, number_of_states, state_probability)
    499 
    500         basis_states = np.arange(number_of_states)
--> 501         return np.random.choice(basis_states, shots, p=state_probability)
    502 
    503     @staticmethod

mtrand.pyx in numpy.random.mtrand.RandomState.choice()
```

**Comment by josh146** [comment]
@kharazity nice catch, this seems to have caught a bug in the `default.mixed` simulator!

In particular, when this line of code is being called with `group[4]`, 
https://github.com/PennyLaneAI/pennylane/blob/16aac9ff7092a94324b27d6b0dfae6f4af43a092/pennylane/_qubit_device.py#L501
the resulting probability distribution is
```
[-9.02056208e-17 -3.28579708e-17  1.38777878e-16  3.47535622e-01
 -5.04087560e-17  4.25007252e-17  1.07365717e-01 -6.93889390e-18
  1.38777878e-17  1.36655339e-01 -3.08148791e-33 -3.71859972e-17
  4.08443321e-01 -1.38777878e-17  1.63693155e-17 -1.38777878e-17]
```

I think we can fix this by ensuring that `default.mixed` takes the absolute value of the probabilities before returning them.

**Comment by josh146** [comment]
@kharazity I've opened a PR, #1680, to fix this latter bug :)
